staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go

```diff
@@ -158,3 +161,92 @@ func ValidateCustomResourceDefinitionNames(names *apiextensions.CustomResourceDe
 
 	return allErrs
 }
+
+// ValidateCustomResourceDefinitionSchema statically validates
+func ValidateCustomResourceDefinitionSchema(JSONSchema *apiextensions.JSONSchemaProps, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	if JSONSchema == nil {
+		return allErrs
+	}
+
+	if JSONSchema.UniqueItems == true {
+		allErrs = append(allErrs, field.Forbidden(fldPath.Child("UniqueItems"), "uniqueItems cannot be set to true since the runtime complexity becomes quadratic"))
+	}
+
+	if JSONSchema.AdditionalProperties != nil {
+		if JSONSchema.AdditionalProperties.Allows == false {
+			allErrs = append(allErrs, field.Forbidden(fldPath.Child("AdditionalProperties"), "additionalProperties cannot be set to false"))
+		}
+		allErrs = append(allErrs, ValidateCustomResourceDefinitionSchema(JSONSchema.AdditionalProperties.Schema, fldPath.Child("AdditionalProperties"))...)
+	}
+
+	if JSONSchema.Ref != "" {
+		openapiRef, err := spec.NewRef(JSONSchema.Ref)
+		if err != nil {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("Ref"), JSONSchema.Ref, err.Error()))
+		}
+		if !openapiRef.IsValidURI() {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("Ref"), JSONSchema.Ref, err.Error()))
+		}
+	}
+
+	if JSONSchema.AdditionalItems != nil {
+		allErrs = append(allErrs, ValidateCustomResourceDefinitionSchema(JSONSchema.AdditionalItems.Schema, fldPath.Child("AdditionalItems"))...)
+	}
+
+	allErrs = append(allErrs, ValidateCustomResourceDefinitionSchema(JSONSchema.Not, fldPath.Child("Not"))...)
+
+	if len(JSONSchema.AllOf) != 0 {
+		for _, jsonSchema := range JSONSchema.AllOf {
+			allErrs = append(allErrs, ValidateCustomResourceDefinitionSchema(&jsonSchema, fldPath.Child("AllOf"))...)
+		}
+	}
+
+	if len(JSONSchema.OneOf) != 0 {
+		for _, jsonSchema := range JSONSchema.OneOf {
+			allErrs = append(allErrs, ValidateCustomResourceDefinitionSchema(&jsonSchema, fldPath.Child("OneOf"))...)
+		}
+	}
+
+	if len(JSONSchema.AnyOf) != 0 {
+		for _, jsonSchema := range JSONSchema.AnyOf {
+			allErrs = append(allErrs, ValidateCustomResourceDefinitionSchema(&jsonSchema, fldPath.Child("AnyOf"))...)
+		}
+	}
+
+	if len(JSONSchema.Properties) != 0 {
+		for property, jsonSchema := range JSONSchema.Properties {
+			allErrs = append(allErrs, ValidateCustomResourceDefinitionSchema(&jsonSchema, fldPath.Child("Properties."+property))...)
```

Comment:
PTAL this please. :)

The fldPath will now be something like `spec.Validation.JSONSchema.Properties.zeta.UniqueItems`. I am not sure if it is a good idea to have the `zeta` part? But it denotes _which property_ the child fields are under.

The other alternative is to have the fldPath like this: `spec.Validation.JSONSchema.Properties.UniqueItems`. This does not tell us the property under which `UniqueItems` lies.